mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,255 @@
+        Degree of spherical harmonic
+    m : int
+        Order of spherical harmonic
+    az : float
+        Azimuthal (longitudinal) spherical coordinate [0, 2*pi]. 0 is aligned
+        with x-axis.
+    pol : float
+        Polar (or colatitudinal) spherical coordinate [0, pi]. 0 is aligned
+        with z-axis.
+
+    Returns
+    -------
+    y : complex float
+        The spherical harmonic value at the specified polar and azimuth angles
+    """
+    assert np.abs(m) <= l, 'Absolute value of expansion coefficient must be <= l'
+
+    # Get function for Legendre derivatives
+    #lut, n_fact = _get_legen_table('meg', False, 100)
+    #lut_deriv_fun = partial(_get_legen_lut_accurate, lut=lut)
+
+    #TODO: Decide on notation for spherical coords
+    #TODO: Should factorial function use floating or long precision?
+
+    #Ensure that polar and azimuth angles are
+    polar = np.array(pol)
+    azimuth = np.array(az)
+
+    # Real valued spherical expansion as given by Wikso, (11)
+    base = np.sqrt(((2 * l + 1) / (4 * np.pi) *
+                    factorial(l - m) / factorial(l + m)) *
+                   lpmv(m, l, np.cos(polar)))
+    if m < 0:
+        return base * np.sin(m * azimuth)
+    else:
+        return base * np.cos(m * azimuth)
+
+    #TODO: Check speed of taking real part of scipy's sph harmonic function
+    # Note reversal in notation order between scipy and original SSS papers
+    # Degree/order and theta/phi reversed
+    #return np.real(sph_harm(m, l, azimuth, polar))
+
+
+def get_num_harmonics(L_in, L_out):
+    """
+    Compute total number of spherical harmonics.
+
+    Parameters
+    ---------
+    L_in : int
+        Internal expansion order
+    L_out : int
+        External expansion order
+
+    Returns
+    -------
+    M : int
+        Total number of spherical harmonics
+    """
+
+    return L_in ** 2 + 2 * L_in + L_out ** 2 + 2 * L_out
+
+
+def _alegendre_deriv(l, m, x):
+    """
+    Compute the derivative of the associated Legendre polynomial at x.
+
+    Parameters
+    ----------
+    l : int
+        Degree of spherical harmonic
+    m : int
+        Order of spherical harmonic
+    x : float
+        Value to evaluate the derivative at
+
+    Returns
+    -------
+    dPlm
+        Associated Legendre function derivative
+    """
+
+    #TODO: Eventually, probably want to switch to look up table but optimize
+    # later
+    C = 1
+    if m < 0:
+        m = abs(m)
+        C = (-1) ** m * fact(l - m) / fact(1 + m)
+    return C * (m * x * lpmv(m, l, x) + (l + m) * (l - m + 1) *
+                np.sqrt(1 - x ** 2) * lpmv(m, l - 1, x)) / (1 - x ** 2)
+
+
+def grad_in_comp(l, m, r, az, pol, lut_func):
+    """
+    Compute gradient of in-component of V(r) spherical expansion having form
+    Ylm(theta, phi) / (r ** (l+1))
+
+    Parameters
+    ----------
+    l : int
+        Degree of spherical harmonic
+    m : int
+        Order of spherical harmonic
+    r : numpy array of shape [n_samples,]
```

Comment:
and I think we've settled on `r : ndarray, shape (n_samples,)` for the way to write this -- is that right @dengemann?